src/Forms/GridField/GridField.php

```diff
@@ -444,6 +451,63 @@ private function initState(): void
         }
     }
 
+    /**
+     * Adds state for this gridfield from the request variables.
+     *
+     * If there is state already set on this GridField, that takes precedent
+     * over state from the request.
+     */
+    private function addStateFromRequest(): void
+    {
+        $request = $this->getRequest();
+        if (($request instanceof NullHTTPRequest) && Controller::has_curr()) {
+            $request = Controller::curr()->getRequest();
+        }
+        // if ($request->params()['Action']) {
+        //     return;
+        // }
+        $stateStr = $this->getStateManager()->getStateFromRequest($this, $request);
+
+        if ($stateStr) {
+            $oldState = $this->getState(false);
+            // Create a dummy state so that we can merge the current state with the request state.
+            $newState = new GridState($this, $stateStr);
+            // Put the current state on top of the request state.
+            $newState->setValue($oldState->Value());
+            $this->state = $newState;
+        }
+    }
+
+    /**
+     * Add get and post parameters if exist to the URL.
+     * 
+     * URL keeps state for GridField, that exist in few different tabs.
+     */
+    public function addRequestToURL(string $link): string
+    {
+        $request = $this->getRequest();
+
+        $postparams = [];
+
+        if ($request->postVars()) {
+            foreach ($request->postVars() as $key => $val) {
+                if (stripos($key, 'gridState') === 0) {
+                    $postparams[$key] = $val;
+                }
+            }
+        }
+
+        $params = array_merge($request->getVars(), $postparams);
```

Comment:
Why are we looking at the GET and POST vars here? Has the state not been stored in the session yet at this point? Compare with `addStateFromRequest()` above where we're getting the state directly from the session.

Reply:
- For this question(https://github.com/silverstripe/silverstripe-framework/pull/10331#issuecomment-1188352227), I've made some changes in GridField.js and all this changes in this branch: https://github.com/silverstripe/silverstripe-admin/pull/1322. If you have a time could you have a look it, please, as well.

- Since, GridField sends GET request for Filtering and POST request for Sorting, we should collect hashes from both Vars. I removed State from URL, I left only hash to access Value in Session.

- We set state as before in this function https://github.com/silverstripe/silverstripe-framework/blob/f5197d5641a1dab316fb46eba142c2f9343de116/src/Forms/GridField/GridFieldStateManager.php#L40

- addStateFromRequest() gets state from Session.

Reply:
As below - I think this can be replaced with the existing functionality.

Reply:
I leave as is. Keep State object in URL. Removed hash.

Reply:
DONE

Reply:
Agreed that this turns out to be necessary - this allows handling nested gridfields, whose state is part of the request and which this gridfield and the state manager aren't explicitly aware of.